Dump devenv into file

xclaesse opened the pull request:
Allow dumping into file and select a format
    
It is often more useful to generate shell script than dumping to stdout.
It is also important to be able to select the shell format.
    
Formats currently implemented:
- sh: Basic VAR=prepend_value:$VAR
- export: Same as 'sh', but also export VAR
- vscode: Same as 'sh', but without substitutions because they don't
  seems to work. To be used in launch.json's envFile.

Dump into builddir/vscode.env file by default

xclaesse commented:
Fox context, there are MRs to improve GStreamer vscode integration, but I think it should be done in Meson itself:
- https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/3719
- https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/1386

bonzini commented:
I am not sure about this. I really would prefer to have the devenv accessible as an environment object, and perhaps you can also add a dump method to environment?

Also should there be support for Windows shells (CMD and PowerShell)?

xclaesse commented:
> I am not sure about this. I really would prefer to have the devenv accessible as an environment object, and perhaps you can also add a dump method to environment?

Not sure what do you mean, how does it help to have env object?

> Also should there be support for Windows shells (CMD and PowerShell)?

Yes, that can be added later.

eli-schwartz commented on mesonbuild/msetup.py:
```diff
@@ -299,6 +300,10 @@ def _finalize_devenv(self, b: build.Build, intr: interpreter.Interpreter) -> Non
             if devenv:
                 b.devenv.append(devenv)
+        # Dump devenv into a file that can be used by vscode's launch.json
+        dump_path = os.path.join(self.build_dir, 'vscode.env')
```
This creates a brand new reserved filename in the build directory root?

Also since apparently this is vscode-specific maybe only create it if vscode is installed.

xclaesse replied:
> This creates a brand new reserved filename in the build directory root?

Right, maybe it should be prefixed with `meson-`.

> Also since apparently this is vscode-specific maybe only create it if vscode is installed.

I'm not sure what other IDE could benefit from this feature, maybe that's more generic than just vscode?

xclaesse replied:
Maybe only generate vscode.env if `.vscode/launch.json` file exists?

eli-schwartz replied:
Sure, maybe. Alternatively it could be created by the vscode extension.

But regardless -- I say this is vscode specific because the filename has the word "vscode" in it, which strikes me as exceedingly specific. ;) If the idea was to be generic then it would have to be, well, generic. :)

xclaesse replied:
> Sure, maybe. Alternatively it could be created by the vscode extension.

That's actually a good point, as long as we have the CLI to dump to file, it can be done by vscode extension indeed. I'm just worried that if user reconfigure the project outside of vscode, the env file might not be up to date. But I guess that's true for every introspection data the extension uses, it could monitor coredata.dat and regenerate automatically.

> But regardless -- I say this is vscode specific because the filename has the word "vscode" in it, which strikes me as exceedingly specific. ;) If the idea was to be generic then it would have to be, well, generic. :)

Right, I named it like that because I only know about vscode that needs it, but if someone knows other IDEs that could use it, then we could have a more generic name. But as said above, I think I'll drop the commit that generate the file automatically and leave that to vscode extension.

xclaesse commented:
> You can do `meson.project_devenv().dump('vscode.env', format: 'vscode')`.

The complete devenv is only known at the end of the configuration. But with https://github.com/mesonbuild/meson/pull/9855 we could let projects do `custom_target(output: 'vscode.env', command: meson.cmd_array() + ['devenv', '--dump', '@OUTPUT@', '--dump-format', 'vscode'])`. But IMHO it doesn't cost much to just always write that file, it's easier.

bonzini commented:
> The complete devenv is only known at the end of the configuration

There are various cases here:

* there could be further calls to add_devenv? In that case `project_devenv()` could forbid other calls to `add_devenv()`, similar to `add_project_arguments()`.

* there's `reduce_winepath` which can either be computed when the environment object is constructed, or left out since `meson test` alredy does it anyway

* finally, things that need access to the install data (gdb, bash completions and the like). would not be included anyway in `--dump` so it's okay to leave them out if you get an environment object for the devenv.

> But IMHO it doesn't cost much to just always write that file, it's easier.

Why would vscode be preferred to other formats, especially since it has some limitations? Some people might prefer a shell file to be sourced, like Python's venv `activate` file, should `meson setup` also create an `activate.{sh,bash,cmd,ps1}` file? I don't think these files are in the same category as `compile_commands.json` for example.

In fact, I wonder if it shouldn't be the task of a vscode extension instead.

Note, I have no objection to adding more formats to `meson devenv --dump`, including the vscode format.

xclaesse commented:
> there could be further calls to add_devenv? In that case project_devenv() could forbid other calls to add_devenv(), similar to add_project_arguments().

Not really, because devenv is global, you want the env from subprojects too.

@bonzini @eli-schwartz I removed the code that dump into a file by default, I think it's better to do that from vscode extension.

Added doc and tests now.

xclaesse commented:
Made a PR in vscode extension to generate `<builddir>/vscode.env`: https://github.com/mesonbuild/vscode-meson/pull/107.

eli-schwartz commented on docs/markdown/snippets/devenv.md:
```diff
@@ -0,0 +1,11 @@
+## Dump devenv into file and select format
+
+`meson devenv --dump [<filename>]` command now takes an option filename argument
```
"now takes an option filename" makes it sound like -- dump always existed and the filename was added.
